# HG changeset patch
# User Rik <rik@octave.org>
# Date 1484002652 28800
#      Mon Jan 09 14:57:32 2017 -0800
# Branch stable
# Node ID 4b5efc5d52a929af734625ae16fb6375ce9f50ae
# Parent  cae62c8d4f96094dd2f6cc1e7d32078d493b4b4c
Calculate error in solution for ode solvers correctly (bug #49950).

* AbsRel_norm.m: When NormControl is on, compare the norm of the solution error
to a tolerance value based on AbsTol, RelTol, and the norm of the solution.
When NormControl is off, compare the absolute value of the solution error to a
tolerance value based on AbsTol, RelTol, and the absolute value of the
solution.

diff --git a/scripts/ode/private/AbsRel_norm.m b/scripts/ode/private/AbsRel_norm.m
--- a/scripts/ode/private/AbsRel_norm.m
+++ b/scripts/ode/private/AbsRel_norm.m
@@ -19,19 +19,18 @@
 
 ## -*- texinfo -*-
 ## @deftypefn {} {retval =} AbsRel_norm (@var{x}, @var{x_old}, @var{AbsTol}, @var{RelTol}, @var{normcoontrol}, @var{y})
 ## Undocumented internal function.
 ## @end deftypefn
 
 function retval = AbsRel_norm (x, x_old, AbsTol, RelTol, normcontrol, y = zeros (size (x)))
 
-  n = numel (x);
-
-  sc = AbsTol + max (abs (x), abs (x_old)) .* RelTol;
   if (normcontrol)
-    retval = max (abs (x - y) ./ sc);
+    sc = max (AbsTol, RelTol * max (sqrt (sumsq (x)), sqrt (sumsq (x_old))));
+    retval = sqrt (sumsq ((x - y))) / sc;
   else
-    retval = sqrt ((1 / n) * sumsq ((x - y) ./ sc));
+    sc = max (AbsTol, RelTol .* max (abs (x), abs (x_old)));
+    retval = max (abs (x - y) ./ sc);
   endif
 
 endfunction
 
